### PR TITLE
Bugfixes, refactoring

### PR DESCRIPTION
**Pull request by czechboy0**
- changed port to just be UInt16 instead of an enum
- fixed wrong endianness when parsing port from ResolvedInternetAddress
- fixed potential leaks in cases when ResolvedInternetAddress was getting created
- added local and remote address getters on sockets
- added a way to wait on a socket to become readable with a timeout

### CODE REVIEW

File: Sources/SocksCore/Address.swift [hunk 1]
```diff
@@ -167,3 +169,32 @@ extension ResolvedInternetAddress: CustomStringConvertible {
...
+        return clientAddress
+    }
+    
+    public func localAddress() throws -> ResolvedInternetAddress {
+        var length = socklen_t(sizeof(sockaddr_storage))
+        let addr = UnsafeMutablePointer<sockaddr_storage>.init(allocatingCapacity: 1)
+        let addrSockAddr = UnsafeMutablePointer<sockaddr>(addr)
+        let res = getsockname(descriptor, addrSockAddr, &length)
+        guard res > -1 else {
+            addr.deallocateCapacity(1)
```
Comment (tannernelson) [inline Sources/SocksCore/Address.swift]: Might be better to do a `defer` directly after the init. That way if someone adds a different early return, there won't be a bug.

```swift
let addr = UnsafeMutablePointer<sockaddr_storage>.init(allocatingCapacity: 1)
defer {
    addr.deallocateCapacity(1)
}
...
guard res > -1 else {
    throw Error(.localAddressResolutionFailed)
}
```

**Comment by tannernelson** [comment]
These look good to me. My only suggestion would be the usage of `defer` over including deallocation calls in the guard else statements.

File: Sources/SocksCore/Address.swift (hunk 1 above)
Comment (czechboy0) [inline Sources/SocksCore/Address.swift]: That's not the same thing. `ResolvedInternetAddress` takes a +1 retained raw pointer, which it just takes over and later deallocates itself. So we need to manually deallocate it here only in the case that `ResolvedInternetAddress` wasn't already created. If I put it in defer, it'd run every time, destroying the memory that `ResolvedInternetAddress` needs. Not too elegant, but I had real hard time to get it all work on both OS X and Linux before I ended up with this solution (there was some stack allocation weirdness on Linux).

**Comment by czechboy0** [comment]
Thanks, see above for a response on why we can't really use `defer`.

File: Sources/SocksCore/Address.swift (hunk 1 above)
Comment (tannernelson) [inline Sources/SocksCore/Address.swift]: Ahhh makes sense. Very cool. Great work.

**Pull request merged**